src/red_black_tree.jl

```diff
@@ -0,0 +1,380 @@
+"""
+function insert_node!(tree::RBTree, node::RBTreeNode)
+    node_y = nothing
+    node_x = tree.root
+
+    while node_x != tree.nil
+        node_y = node_x
+        if node.data < node_x.data
+            node_x = node_x.leftChild
+        else
+            node_x = node_x.rightChild
+        end
+    end
+
+    node.parent = node_y
+    if node_y == nothing
+        tree.root = node
+    elseif node.data < node_y.data
+        node_y.leftChild = node
+    else
+        node_y.rightChild = node
+    end
+end
+
+"""
+    left_rotate!(tree::RBTree, node_x::RBTreeNode)
+
+Performs a left-rotation on `node_x` and updates `tree.root`, if required.
+"""
+function left_rotate!(tree::RBTree, node_x::RBTreeNode)
+    node_y = node_x.rightChild
+    node_x.rightChild = node_y.leftChild
+    if node_y.leftChild != tree.nil
+        node_y.leftChild.parent = node_x
+    end
+    node_y.parent = node_x.parent
+    if (node_x.parent == nothing)
+        tree.root = node_y
+    elseif (node_x == node_x.parent.leftChild)
+        node_x.parent.leftChild = node_y
+    else
+        node_x.parent.rightChild = node_y
+    end
+    node_y.leftChild = node_x
+    node_x.parent = node_y
+end    
+
+"""
+    right_rotate!(tree::RBTree, node_x::RBTreeNode)
+
+Performs a right-rotation on `node_x` and updates `tree.root`, if required.
+"""
+function right_rotate!(tree::RBTree, node_x::RBTreeNode)
+    node_y = node_x.leftChild
+    node_x.leftChild = node_y.rightChild
+    if node_y.rightChild != tree.nil
+        node_y.rightChild.parent = node_x
+    end
+    node_y.parent = node_x.parent
+    if (node_x.parent == nothing)
+        tree.root = node_y
+    elseif (node_x == node_x.parent.leftChild)
+        node_x.parent.leftChild = node_y
+    else
+        node_x.parent.rightChild = node_y
+    end
+    node_y.rightChild = node_x
+    node_x.parent = node_y
+end    
+
+"""
+   fix_insert!(tree::RBTree, node::RBTreeNode) 
+
+This method is called to fix the property of having no two adjacent nodes of red color in the `tree`.
+"""
+function fix_insert!(tree::RBTree, node::RBTreeNode)
+    parent = nothing
+    grand_parent = nothing
+    # for root node, we need to change the color to black
+    # other nodes, we need to maintain the property such that
+    # no two adjacent nodes are red in color
+    while  node != tree.root && node.parent.color
+        parent = node.parent
+        grand_parent = parent.parent
+        
+        if (parent == grand_parent.leftChild) # parent is the leftChild of grand_parent
+            uncle = grand_parent.rightChild
+
+            if (uncle.color) # uncle is red in color
+                grand_parent.color = true
+                parent.color = false
+                uncle.color = false
+                node = grand_parent
+            else  # uncle is black in color
+                if (node == parent.rightChild) # node is rightChild of it's parent
```

Comment:
```suggestion
                if (node == parent.rightChild) # node is rightChild of its parent
```

https://www.grammarly.com/blog/its-vs-its/